src/weakptr.c

```diff
@@ -697,14 +714,24 @@ void MakeImmutableWPObj( Obj obj )
   SET_LEN_PLIST(copy, len);
   SET_PTR_BAG(obj, PTR_BAG(copy));
 #else
-  /* remove any weak dead bags */
-  for (i = 1; i <= STORED_LEN_WPOBJ(obj); i++) {
-      Obj elm = ELM_WPOBJ(obj,i);
-      if (elm != 0 && IS_WEAK_DEAD_BAG(elm))
-        ELM_WPOBJ(obj,i) = 0;
-  }
-  /* Change the type */
-  RetypeBag( obj, T_PLIST+IMMUTABLE);
+    // recompute stored length
+    UInt len = LengthWPObj(obj);
+
+    // remove any weak dead bags, by relying on side-effect of ELM_WPOBJ
+    for (UInt i = 1; i <= len; i++) {
+        Obj elm = ELM_WPOBJ(obj, i);
+        if (IS_BAG_REF(elm)) {
+            MakeImmutable(elm);
+        }
+    }
+
+    // change the type - this works correctly, as the layout of weak pointer
```

Comment:
Do we need to recompute the length here? If a GC somehow managed to happen between the line 718 and here, then the last item in the weak pointer object might have evaporated, and `len` will now be too large. So the question is whether `MakeImmutable` can cause a garbage collection, which I think it probably can, thanks to `PostMakeImmutable` calls.

Reply:
You are right. And it's worse: if a GC happens, then it may mark objects we already touched as halfdead (with GASMAN), or crash (in Julia GC, because we now have a partially invalid object).

One way to fix this is to first convert to a mutable plist (which causes no GC), then make that immutable). I'll look into it.